src/Native/System.Security.Cryptography.Native/pal_ssl.cpp

```diff
@@ -41,6 +43,256 @@ extern "C" SSL_CTX* SslCtxCreate(SSL_METHOD* method)
+            return CipherAlgorithmType::Aes128;
+
+        case SSL_CipherAlgorithm::SSL_AES256GCM:
+            return CipherAlgorithmType::Aes256;
+#endif
+    }
+
+    return CipherAlgorithmType::None;
+}
+
+static ExchangeAlgorithmType MapExchangeAlgorithmType(const SSL_CIPHER* cipher)
+{
+    unsigned long mkey;
+#ifdef HAVE_SSL_CIPHER_ALGORITHMS
+    const unsigned long SSL_MKEY_MASK = 0x000000FFL;
+    mkey = cipher->algorithms & SSL_MKEY_MASK;
+#else
+    mkey = cipher->algorithm_mkey;
+#endif
+
+    SSL_KeyExchangeAlgorithm sslMkey = static_cast<SSL_KeyExchangeAlgorithm>(mkey);
+    switch (sslMkey)
+    {
+        case SSL_KeyExchangeAlgorithm::SSL_kRSA:
+            return ExchangeAlgorithmType::RsaKeyX;
+
+        case SSL_KeyExchangeAlgorithm::SSL_kDHr:
+            return ExchangeAlgorithmType::DiffieHellman;
+
+        case SSL_KeyExchangeAlgorithm::SSL_kDHd:
+            return ExchangeAlgorithmType::DiffieHellman;
+
+        case SSL_KeyExchangeAlgorithm::SSL_kEDH:
+            return ExchangeAlgorithmType::DiffieHellman;
+
+        case SSL_KeyExchangeAlgorithm::SSL_kKRB5:
+            return ExchangeAlgorithmType::SSL_kKRB5;
+
+#ifdef HAVE_SSL_CIPHER_ALGORITHMS
+        case SSL_KeyExchangeAlgorithm::SSL_kECDH:
+            return ExchangeAlgorithmType::SSL_ECDH;
+
+        case SSL_KeyExchangeAlgorithm::SSL_kECDHE:
+            return ExchangeAlgorithmType::SSL_ECDSA;
+#else
+        case SSL_KeyExchangeAlgorithm::SSL_kECDHr:
+            return ExchangeAlgorithmType::SSL_ECDH;
+
+        case SSL_KeyExchangeAlgorithm::SSL_kECDHe:
+            return ExchangeAlgorithmType::SSL_ECDSA;
+
+        case SSL_KeyExchangeAlgorithm::SSL_kEECDH:
+            return ExchangeAlgorithmType::SSL_ECDSA;
+
+        case SSL_KeyExchangeAlgorithm::SSL_kPSK:
+            return ExchangeAlgorithmType::SSL_kPSK;
+
+        case SSL_KeyExchangeAlgorithm::SSL_kGOST:
+            return ExchangeAlgorithmType::SSL_kGOST;
+
+        case SSL_KeyExchangeAlgorithm::SSL_kSRP:
+            return ExchangeAlgorithmType::SSL_kSRP;
+#endif
+    }
+
+    return ExchangeAlgorithmType::None;
+}
+
+static HashAlgorithmType MapHashAlgorithmType(const SSL_CIPHER* cipher)
+{
+    unsigned long mac;
+#ifdef HAVE_SSL_CIPHER_ALGORITHMS
+    const unsigned long SSL_MAC_MASK = 0x00c00000L;
+    mac = cipher->algorithms & SSL_MAC_MASK;
+#else
+    mac = cipher->algorithm_mac;
+#endif
+
+    SSL_DataHashAlgorithm sslMac = static_cast<SSL_DataHashAlgorithm>(mac);
+    switch (sslMac)
+    {
+        case SSL_DataHashAlgorithm::SSL_MD5:
+            return HashAlgorithmType::Md5;
+
+        case SSL_DataHashAlgorithm::SSL_SHA1:
+            return HashAlgorithmType::Sha1;
+
+#ifndef HAVE_SSL_CIPHER_ALGORITHMS
+        case SSL_DataHashAlgorithm::SSL_GOST94:
+            return HashAlgorithmType::SSL_GOST94;
+
+        case SSL_DataHashAlgorithm::SSL_GOST89MAC:
+            return HashAlgorithmType::SSL_GOST89;
+
+        case SSL_DataHashAlgorithm::SSL_SHA256:
```

Comment:
Since this is under ifndef, does it mean that OpenSsl 1.0.* doesn't support SHA256, SHA384 and AEAD?